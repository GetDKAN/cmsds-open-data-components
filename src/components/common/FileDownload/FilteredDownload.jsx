import React from 'react'
import PropTypes from 'prop-types'
import FontAwesomePro from '../FontAwesomePro/FontAwesomePro'
import DatasetDownloadLink from '../../dataset/DatasetDownloadLink/DatasetDownloadLink'

const FilteredDownload = ({ resource, datasetTitle, filteredURL, isDatasetExplorerDownloadLink = false }) => {
  let format = null

  if (
    Object.prototype.hasOwnProperty.call(resource, 'data') &&
    Object.prototype.hasOwnProperty.call(resource.data, 'format')
  ) {
    format = resource.data.format
    format.toUpperCase()
  }

  let title = null

  if (
    Object.prototype.hasOwnProperty.call(resource, 'data') &&
    Object.prototype.hasOwnProperty.call(resource.data, 'data') &&
    Object.prototype.hasOwnProperty.call(resource.data.data, 'title')
  ) {
    title = resource.data.data.title
  } else {
    title = 'This Dataset'
  }

  const item = (
    <>
      {filteredURL && title && format && (
        <DatasetDownloadLink
          linkText={(
            <>
              {isDatasetExplorerDownloadLink ? (
                <FontAwesomePro icon='download' aria-hidden />
              ) : (
                <FontAwesomePro icon='filter' aria-hidden />
              )}
              {isDatasetExplorerDownloadLink ? (
                <span>Download filtered dataset ({format.toUpperCase()})</span>
              ) : (
                <span>Download filtered dataset <span className='sr-only'>- {format}</span></span>
              )}
            </>
          )}
          linkHref={filteredURL}
        />
      )}
      {!isDatasetExplorerDownloadLink && (
        <div className='format'>
          <span>
            {format}
          </span>
        </div>
      )}
    </>
  )

  return isDatasetExplorerDownloadLink ? item : <div className='resource'>{item}</div>
}

FilteredDownload.propTypes = {
  /**
   * Dataset data object
   */
  resource: PropTypes.shape({
    data: PropTypes.object,
    format: PropTypes.string,
    identifier: PropTypes.string
  }),
  /**
   * Dataset title
   */
  datasetTitle: PropTypes.string,
  /**
   * Filtered dataset download url
   */
  filteredURL: PropTypes.string,
  /**
   * Dataset explorer download link gets different styling
   */
  isDatasetExplorerDownloadLink: PropTypes.bool
}
FilteredDownload.displayName = 'FilteredDownload'
export default FilteredDownload
