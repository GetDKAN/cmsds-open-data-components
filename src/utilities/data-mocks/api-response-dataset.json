{
    "fixture": true,
    "accessLevel": "public",
    "landingPage": "https://data.cms.gov/provider-data/dataset/test-data-abcd-0005",
    "bureauCode": [
        "009:38"
    ],
    "issued": "2020-01-20",
    "@type": "dcat:Dataset",
    "modified": "2020-01-20",
    "released": "2020-09-17",
    "archiveExclude": false,
    "keyword": [
        {
            "identifier": "8a0cf0a8-a3d0-5223-b1cc-82fbb3101cdb",
            "data": "hhc"
        },
        {
            "identifier": "9929edf6-e8af-5e74-b28e-6c43c0e0a534",
            "data": "general information"
        }
    ],
    "contactPoint": {
        "@type": "vcard:Contact",
        "fn": "Home Health Compare",
        "hasURL": "https://www.cms.gov/apps/contacts"
    },
    "publisher": {
        "identifier": "4c1de1c4-f083-5877-8968-7ddc04756c49",
        "data": {
            "@type": "org:Organization",
            "name": "Centers for Medicare & Medicaid Services (CMS)"
        }
    },
    "identifier": "test-data-abcd-0005",
    "description": "Home Health Care - Measure Date Range",
    "title": "Home Health Care - Measure Date Range",
    "programCode": [
        "009:000"
    ],
    "distribution": [
        {
            "identifier": "716256f0-6d17-56a6-9b5d-c35879011e9e",
            "data": {
                "@type": "dcat:Distribution",
                "downloadURL": "https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv",
                "mediaType": "text/csv",
                "%Ref:downloadURL": [
                    {
                        "identifier": "2850c4af459ff447d4e9bada0637ec54__1701355540__source",
                        "data": {
                            "filePath": "s3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0005.csv",
                            "identifier": "2850c4af459ff447d4e9bada0637ec54",
                            "mimeType": "text/csv",
                            "perspective": "source",
                            "version": "1701355540",
                            "checksum": null
                        }
                    },
                    {
                        "identifier": "2850c4af459ff447d4e9bada0637ec54__1701355540__local_url",
                        "data": {
                            "filePath": "https://h-o.st/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv",
                            "identifier": "2850c4af459ff447d4e9bada0637ec54",
                            "mimeType": "text/csv",
                            "perspective": "local_url",
                            "version": "1701355540",
                            "checksum": null
                        }
                    }
                ]
            },
            "format": "csv"
        }
    ],
    "theme": [
        {
            "identifier": "eda174d2-96df-5537-868f-f5ed625c3ede",
            "data": "Home health services"
        }
    ],
    "%modified": "2023-11-30T09:45:33-0500"
}
