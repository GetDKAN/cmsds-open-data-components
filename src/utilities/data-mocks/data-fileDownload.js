export const resourceData = {
  data: {
    identifier: '11745687-dc47-52ac-b9e4-66c50fe772ad',
    data: {
      '@type': 'dcat:Distribution',
      title: 'Test Data abcd 0016',
      downloadURL: 'https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/f69b073f217ea35ed14997178f7085c8_1677006237/Test_Data_abcd_0016.csv',
      mediaType: 'text/csv',
      '%Ref:downloadURL': [
        {
          identifier: 'f69b073f217ea35ed14997178f7085c8__1677006237__source',
          data: {
            filePath: 's3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0016.csv',
            identifier: 'f69b073f217ea35ed14997178f7085c8',
            mimeType: 'text/csv',
            perspective: 'source',
            version: '1677006237',
            checksum: null
          }
        },
        {
          identifier: 'f69b073f217ea35ed14997178f7085c8__1677006237__local_url',
          data: {
            filePath: 'http://h-o.st/provider-data/sites/default/files/resources/f69b073f217ea35ed14997178f7085c8_1677006237/Test_Data_abcd_0016.csv',
            identifier: 'f69b073f217ea35ed14997178f7085c8',
            mimeType: 'text/csv',
            perspective: 'local_url',
            version: '1677006237',
            checksum: null
          }
        }
      ]
    },
    format: 'csv'
  }
}

export const responseHeaders = {
  'content-length': 99999
}
