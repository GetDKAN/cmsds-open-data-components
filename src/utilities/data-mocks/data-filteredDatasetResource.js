export const filteredDatasetResource = {
  // identifier: '',
  // identifier: '07043497-a5ec-53f0-8a78-230dcd7f3e33',
  data: {
    '@type': 'dcat:Distribution',
    downloadURL: 'https://pqdc-dkan.ddev.site/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv',
    mediaType: 'text/csv',
    '%Ref:downloadURL': [
      {
        identifier: '2850c4af459ff447d4e9bada0637ec54__1701355540__source',
        data: {
          filePath: 's3://913461122956-pdc-dev-test-minimal-data/Test_Data_abcd_0005.csv',
          identifier: '2850c4af459ff447d4e9bada0637ec54',
          mimeType: 'text/csv',
          perspective: 'source',
          version: '1701355540',
          checksum: null
        }
      },
      {
        identifier: '2850c4af459ff447d4e9bada0637ec54__1701355540__local_url',
        data: {
          filePath: 'https://h-o.st/provider-data/sites/default/files/resources/2850c4af459ff447d4e9bada0637ec54_1701355540/Test_Data_abcd_0005.csv',
          identifier: '2850c4af459ff447d4e9bada0637ec54',
          mimeType: 'text/csv',
          perspective: 'local_url',
          version: '1701355540',
          checksum: null
        }
      }
    ]
  },
  format: 'csv'
}
