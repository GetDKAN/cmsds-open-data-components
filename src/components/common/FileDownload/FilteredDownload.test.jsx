import { render, screen } from '@testing-library/react'
import FilteredDownload from './FilteredDownload'
import { resourceData } from '../../../utilities/data-mocks/data-fileDownload'
import { MemoryRouter } from 'react-router-dom'

describe('FilteredDownload component.', () => {
  it('Matches snapshot.', () => {
    const renderedFilteredDownload = render(
      <MemoryRouter
        initialEntries={[{
          pathname: '/',
          search: '',
          hash: '',
          state: {
            fromTitle: '',
            fromUrl: ''
          },
          key: 'abc123'
        }]}
      >
        <FilteredDownload
          resource={resourceData}
          datasetTitle='Home Health Care - Measure Date Range'
          filteredURL='https://pqdc-dkan.ddev.site/provider-data/api/1/pdc/query/test-data-abcd-0005/0/download?conditions%5B0%5D%5Bproperty%5D=col5&conditions%5B0%5D%5Boperator%5D=is_empty&conditions%5B0%5D%5Bvalue%5D=&properties=1-2-3-4-5-6&format=csv'
        />
      </MemoryRouter>
    )

    expect(renderedFilteredDownload.asFragment()).toMatchSnapshot()
  })

  it('Does not render download link if resource prop data is not provided.', () => {
    render(
      <MemoryRouter
        initialEntries={[{
          pathname: '/',
          search: '',
          hash: '',
          state: {
            fromTitle: '',
            fromUrl: ''
          },
          key: 'abc123'
        }]}
      >
        <FilteredDownload resource={{}} />
      </MemoryRouter>
    )

    expect(screen.queryByRole('link')).not.toBeInTheDocument()
  })
})
