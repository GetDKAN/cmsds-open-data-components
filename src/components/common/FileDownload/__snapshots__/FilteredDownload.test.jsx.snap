// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilteredDownload component. > Matches snapshot. 1`] = `
<DocumentFragment>
  <div
    class="resource"
  >
    <a
      class="dataset-download-link"
      href=""
      rel="noopener noreferrer"
      role="button"
      tabindex="0"
      target="_blank"
    >
      <svg
        aria-hidden="true"
        aria-label="filter"
        class="fa fa-filter"
        focusable="false"
        height="16"
        role="img"
        viewBox="0 0 512 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M463.952 0H48.057C5.419 0-16.094 51.731 14.116 81.941L176 243.882V416c0 15.108 7.113 29.335 19.2 40l64 47.066c31.273 21.855 76.8 1.538 76.8-38.4V243.882L497.893 81.941C528.042 51.792 506.675 0 463.952 0zM288 224v240l-64-48V224L48 48h416L288 224z"
          fill="#323A45"
        />
      </svg>
      <span>
        Download filtered dataset 
        <span
          class="sr-only"
        >
          - csv
        </span>
      </span>
    </a>
    <div
      class="ds-c-dialog-wrap"
    >
      <dialog
        aria-labelledby="dialog--1__heading"
        class="ds-c-dialog csv-download-link-dialog"
        id="dialog--1"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ds-c-dialog__window"
        >
          <div
            class="ds-c-dialog__header"
          >
            <h2
              class="ds-c-dialog__heading"
              id="dialog--1__heading"
            >
              Notice: Downloading in Excel
            </h2>
            <button
              aria-label="Close notice dialog"
              class="ds-c-close-button ds-c-dialog__close"
              id="dialog--1__close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin "
                id="dialog--1__close__icon"
                viewBox="-2 -2 18 18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 13.0332964L13.0332964 0M13.0332964 13.0332964L0 0"
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="ds-c-dialog__body"
          >
            <div
              id="dialog--1__content"
            >
              <p>
                Before downloading these files in Excel, please review the 
                <strong>
                  Excel Download Instructions
                </strong>
                 to ensure accuracy with the display of your data when downloading, opening and saving these files.
              </p>
              <ul
                class="csv-faq-link"
              >
                <li>
                  <a
                    href="/about#download-files-in-excel"
                  >
                    Excel Download Instructions
                  </a>
                </li>
              </ul>
              <p>
                Would you like to proceed?
              </p>
            </div>
            <div
              class="ds-c-dialog__actions"
            >
              <div
                class="csv-download-link-dialog-actions"
              >
                <button
                  class="btn blueberry"
                  data-download-url="https://pqdc-dkan.ddev.site/provider-data/api/1/pdc/query/test-data-abcd-0005/0/download?conditions%5B0%5D%5Bproperty%5D=col5&conditions%5B0%5D%5Boperator%5D=is_empty&conditions%5B0%5D%5Bvalue%5D=&properties=1-2-3-4-5-6&format=csv"
                >
                  Yes, download
                </button>
                <button
                  class="btn"
                >
                  No, cancel
                </button>
              </div>
            </div>
          </div>
        </div>
      </dialog>
    </div>
    <div
      class="format"
    >
      <span>
        csv
      </span>
    </div>
  </div>
</DocumentFragment>
`;
